app/classes/observation_report/darwin/images.rb

```diff
@@ -0,0 +1,103 @@
+# frozen_string_literal: true
+
+module ObservationReport
+  module Darwin
+    # Darwin Core Observations format.
+    class Images < ObservationReport::CSV
+      attr_accessor :observations, :query, :tables
+
+      self.separator = "\t"
+
+      def initialize(args)
+        super(args)
+        self.observations = args[:observations]
+        initialize_query
+      end
+
+      def initialize_query
+        add_joins
+        add_project
+      end
+
+      def add_joins
+        self.tables = {
+          images: Image.arel_table,
+          images_observations: Arel::Table.new(:images_observations),
+          observations: Observation.arel_table,
+          users: User.arel_table,
+          licenses: License.arel_table
+        }
+        self.query = tables[:images]
+        join_table(:images_observations, :image_id, attribute(:images, :id))
+        io_attribute = attribute(:images_observations, :observation_id)
+        join_table(:observations, :id, io_attribute)
+        join_table(:users, :id, attribute(:images, :user_id))
+        join_table(:licenses, :id, attribute(:images, :license_id))
+      end
+
+      def join_table(join_name, join_field, attribute)
+        table = tables[join_name]
+        join_attribute = table[join_field]
+        self.query = query.join(table).on(join_attribute.eq(attribute))
+      end
+
+      def add_project
+        query.project(attribute(:images, :id),
+                      attribute(:observations, :id).as("obs_id"),
+                      attribute(:images, :when),
+                      attribute(:users, :name),
+                      attribute(:users, :login),
+                      attribute(:licenses, :url).as("license_url"),
+                      attribute(:images, :copyright_holder))
+      end
+
+      def attribute(table_name, field)
+        tables[table_name][field]
+      end
+
+      def formatted_rows
+        query.where(tables[:observations][:id].in(observations.ids))
+        rows = ActiveRecord::Base.connection.exec_query(query.to_sql)
+        sort_after(rows.map { |row| format_image_row(row) })
+      end
+
+      def labels
+        %w[
+          identifier
+          type
+          format
+          accessURL
+          created
+          creator
+          license
+          rightsHolder
+        ]
+      end
+
+      def format_row(row)
+        row
+      end
+
+      def sort_after(rows)
+        rows.sort_by { |row| row[0].to_i }
+      end
+
+      private
+
+      def image_url(id)
+        "https://images.mushroomobserver.org/1280/#{id}.jpg"
```

Comment:
The "correct" URL is https://mushroomobserver.org/images/$size/$id.jpg.  For 1280 that is currently identical, but we cannot guarantee it long-term.  Also, this corrected URL will automatically server the image locally if it has not been transferred to the image server yet.  Yes, it generally only takes a few seconds to transfer images, but on rare occasions owing to glitches, etc., it can take up to a day before the nightly script/verify_images job catches any missed transfers.

Reply:
Will do.  I was just worried about putting that much pressure on mushroomobserver.org since I expect folks will grab the DwCA and then immediately start downloading everything.  I wonder if there would be a way to create a tarball somewhere that folks could download instead of hitting all though URLs.  I was also figuring that we'd be generating these things on a slow enough time scale that I wouldn't have to worry about a lag of even a day.  I think the current idea is to generate one of these something like once a quarter.  I'm also debating what the image size should be.  I think Alan Celestino used 320s.  That seems a bit small these days, but maybe that's preferred.  I'm hoping to get Jess and Alan's opinions on that.